taskvine/src/manager/vine_schedule.c

```diff
@@ -209,8 +209,15 @@ int check_worker_against_task(struct vine_manager *q, struct vine_worker_info *w
 
 	/* do this check last! otherwise the library is sent to workers that can't fit it. */
 	if (t->needs_library) {
-		struct vine_task *library = vine_schedule_find_library(w, t->needs_library);
-		if (!library) {
+		int status;
+		struct vine_task *library = vine_schedule_find_library(w, t->needs_library, &status);
+		if (status == LIBRARY_TASK_FOUND_WITH_AVAILABLE_SLOT) {
+			/* If the library is already running on this worker, then it's ok. */
+			return 1;
+		} else if (status == LIBRARY_TASK_FOUND_NO_AVAILABLE_SLOT) {
```

Comment:
If the library has no slots it should do the same as library not found. The same library may run more than once in a worker.

Reply:
Just be confused. In this case, for example, if a user installs 1 library with 16 cores and 16 slots, and the total cores are 64, how many concurrent tasks are expected to be?

Reply:
At least 16 for that library. However the worker would fit 3 more libraries, so it 4 libraries are installed, then 64. That is 64 slots across 4 libraries.

Reply:
Oh, I thought that was an issue... As my experience, sometimes I want to over-allocate the cores on a worker, so that I can easily modify the usable library cores without removing and collecting a new worker. I think sometimes the users may want to have the option to not to use out all available cores on a worker, that was the intention of this pr. What do you think?

Reply:
We want to precisely account for the resources that each task is using. Instead, I suggest having libraries that fill the whole worker, and add an option to the function calls that specifies how many slots they use.  Such an option may be needed anyway as functions from a library may have different resources needs.

Reply:
So for a 64-core worker, this library setting:
``` Python
libtask.set_cores(1)
libtask.set_function_slots(1)
```
has the same effect with:
``` Python
libtask.set_cores(64)
libtask.set_function_slots(64)
```
except that the first one has only 1 single process but the second one has 64 independent processes ... If we would like to fill the worker with libraries, `set_cores` seems to not really matter. How about automatically setting the library cores to the maximum of total cores on the worker? 

Or, can we also encourage the users to leverage as many cores as they can, but also leaves them an option to use partial cores on the worker?

Reply:
It does not have the same effect... One has one library with 64 slots, and the other has 64 libraries, each with one slot. That overhead may become important.

```
libtask.set_cores(60)
libtask.set_function_slots(60)
```
This would be 61 threads? 60 slots + the thread of the library?

Reply:
A user might want to run one concurrent function per library and just take advantage of the fact that the interpreter is not started every time. Also, they may want to run different types of libraries, so a library should not take the whole worker.

Reply:
Given the above, is it ok to close this pr?  We need to allow multiple libraries and library types in a worker, and this pr seems to disallow that.  This is something I often get confused and I forget that a worker may run more than one library.